File: leetcode/go/src/leetcode/109_convert_sorted_list_to_binary_search_tree.go
package leetcode

func sortedListToBSTn(head *ListNode, n int) *TreeNode {
	if n == 0 {
		return nil
	}
	if n == 1 {
		return &TreeNode{Val: head.Val, Left: nil, Right: nil}
	}
	var leftPart *ListNode = head
	var rightPart *ListNode = nil
	for i := 0; i < n/2-1; i++ {
		leftPart = leftPart.Next
	}
	root := &TreeNode{Val: leftPart.Next.Val}
	rightPart = leftPart.Next.Next
	leftPart.Next = nil
	root.Left = sortedListToBSTn(head, n/2)
	if rightPart == nil {
		root.Right = sortedListToBSTn(rightPart, 0)
	} else {
		root.Right = sortedListToBSTn(rightPart, n-n/2-1)
	}
	return root
}

func sortedListToBST(head *ListNode) *TreeNode {
	n := 0
	t := head
	for t != nil {
		n++
		t = t.Next
	}
	return sortedListToBSTn(head, n)
}

File: leetcode/cpp/109_convert_sorted_list_to_binary_search_tree.cpp
class Solution {
 public:
  // time: O(logn*n), space: O(n)
  TreeNode *sortedListToBST(ListNode *head) {
    int c = 0;
    auto root = head;
    while (root != nullptr) {
      c++;
      root = root->next;
    }
    return innerSort(head, c);
  }

  TreeNode *innerSort(ListNode *head, int len) {
    if (len == 0) return nullptr;
    if (len == 1) return new TreeNode(head->val);

    int middle = len / 2;
    ListNode *rroot;
    ListNode *pre = head;
    for (int i = 0; i < middle - 1; i++) {
      pre = pre->next;
    }
    rroot = pre->next->next;
    TreeNode *root = new TreeNode(pre->next->val);
    pre->next = nullptr;

    root->left = innerSort(head, middle);
    root->right = innerSort(rroot, len - middle - 1);
    return root;
  }
};
